### PR TITLE
implement user and login, closes #4 #5

### LINKED ISSUE #4: User
- [ ] `userController`
- [ ] `userModel`

### LINKED ISSUE #5: Log In / Sign Up Modal
build modal for log in sign up after completion of #3

### PR DESCRIPTION
**Pull request by tiffanygan**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Need get rid of login and register options once logged in (replace with logout, possibly settings?)
```
Comment (ganneo) [inline src/index.js]: Remove this line.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import IndexUI from '../src/ui/IndexUI';
+
+const ui = new IndexUI(document);
+let currUser = JSON.parse(localStorage.getItem('currUser'));
```
Comment (ganneo) [inline src/index.js]: Set the key as a global variable and other js can import it from `index.js`

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+window.addEventListener('DOMContentLoaded', () => {
+    if (!currUser) {
+        return;
+    }
+    ui.initPage(currUser.username);
```
Comment (ganneo) [inline src/index.js]: LocalStorage only saves id.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+});
+
+ui.logoutBtn.addEventListener('click', () => {
+    ui.logout();
+    currUser = null;
+    localStorage.setItem('currUser', JSON.stringify(currUser));
```
Comment (ganneo) [inline src/index.js]: LocalStorage only saves id.

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+});
```
Comment (ganneo) [inline src/index.js]: Add empty new line.

File: src/login.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  const checkStatus = await userService.checkUser(ui.email.value, ui.password.value);
+  if (checkStatus === NO_ACCOUNT) {
+    ui.failure("User does not exist");
+  } else if (checkStatus === PASSWORD_WRONG) {
+    ui.failure("Password is wrong");
+  } else {
+    const user = await userService.getUserByEmail(ui.email.value);
+    const id = user.id;
+    ui.clearInputs();
+    localStorage.setItem('currUser', JSON.stringify(user));
```
Comment (ganneo) [inline src/login.js]: LocalStorage only saves id.

File: src/login.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    // See if better way (in case of changing localhost to somthing else)
+    window.location.assign('http://localhost:8080/index.html');
```
Comment (ganneo) [inline src/login.js]: In HTML use a link instead of redirect.

File: src/login.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  }
+});
```
Comment (ganneo) [inline src/login.js]: Add an empty new line.

File: src/model/User.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    constructor(username, email, password) {
+        this.username = username;
+        this.email = email;
+        this.password = password;
+    }
+
+    setId(id) {
+        this.id = id;
+    }
+}
```
Comment (ganneo) [inline src/model/User.js]: Add an empty new line.

File: src/register.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    if (!(await userService.checkUsername(user))) {
+        ui.showUsernameAlert();
+        return;
+    }
+
+    const createdUser = await userService.addUser(user);
+    if (createdUser) {
+        ui.clearInputs();
+        const id = createdUser.id;
+        localStorage.setItem('currUser', JSON.stringify(createdUser));
```
Comment (ganneo) [inline src/register.js]: LocalStorage only saves id.

File: src/register.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        // See if better way (in case of changing localhost to somthing else)
+        window.location.assign('http://localhost:8080/index.html');
```
Comment (ganneo) [inline src/register.js]: In HTML add a link instead of redirect.

File: src/service/UserService.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+// Need md5
```
Comment (ganneo) [inline src/service/UserService.js]: Npm install md5 and import md5.

File: src/service/UserService.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+export const NO_ACCOUNT = 2;
+export const LOGIN_SUCCESS = 3;
+export default class UserService {
+  constructor() {
+    this.userClient = new EasyHttp("http://localhost:3000/users");
+  }
+
+  async checkEmail(user) {
+    const users = await this.userClient.get();
+    if (users.map((currUser) => currUser.email).includes(user.email)) {
```
Comment (ganneo) [inline src/service/UserService.js]: 1. Directly return boolean
2. Change `map` to `some`

File: src/service/UserService.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+      return false;
+    }
+    return true;
+  }
+
+  async checkUsername(user) {
+    const users = await this.userClient.get();
+    if (
+      users.map((currUser) => currUser.username).includes(user.username)
```
Comment (ganneo) [inline src/service/UserService.js]: Same as line 14.

File: src/settings.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
+// need md5
```
Comment (ganneo) [inline src/settings.js]: Install npm md5.

File: src/settings.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    const newUser = new User(ui.newUsername.value, currUser.email, ui.newPassword.value);
+    if (!userService.checkUsername(newUser)) {
+        return;
+    }
+    currUser = await userService.updateUser(currUser.id, newUser);
+    localStorage.setItem('currUser', JSON.stringify(currUser));
+
+    // See if better way (in case of changing localhost to somthing else)
+    window.location.assign('http://localhost:8080/index.html');
+});
```
Comment (ganneo) [inline src/settings.js]: Add new empty line.

File: src/ui/LoginUI.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+        this.failureAlert.style.display = 'block';
+        this.failureMessage.textContent = message;
+        setTimeout(() => this.failureAlert.style.display = 'none', 3000);
+    }
+
+    clearInputs() {
+        this.email.value = '';
+        this.password.value = '';
+    }
+}
```
Comment (ganneo) [inline src/ui/LoginUI.js]: Add new empty line.

File: src/ui/RegisterUI.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        this.confirmPassword.value = '';
+    }
+
+    createUser() {
+        if (this.checkPasswords()) {
+            return new User(this.username.value, this.email.value, this.password.value);
+        }
+        return null;
+    }
+}
```
Comment (ganneo) [inline src/ui/RegisterUI.js]: Add new empty line

File: src/ui/SettingsUI.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+        this.newPassword = this.document.getElementById('new-password');
+        this.saveChangesBtn = this.document.getElementById('save-changes-btn');
+        this.passwordAlert = this.document.getElementById('settings-password-alert');
+    }
+
+    showPasswordAlert() {
+        this.passwordAlert.style.display = 'block';
+        setTimeout(() => this.passwordAlert.style.display = 'none', 2000);
+    }
+}
```
Comment (ganneo) [inline src/ui/SettingsUI.js]: Add new empty line

**Review by ganneo — CHANGES_REQUESTED** [review]

File: src/service/UserService.js [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  async updateUser(id, newUser) {
+    return await this.userClient.update(id, newUser);
+  }
+
+  async getUserByEmail(email) {
+    const users = await this.userClient.get();
+    return users.find((currUser) => currUser.email === email);
+  }
```
Comment (yugan2005) [inline src/service/UserService.js]: For these three methods, their return are simply Object, not the `User` type Object.
Need use `Object.assign(new Foo, jsonPlainObject)` to convert it.
[reference](https://stackoverflow.com/a/43977474/4229125)

File: src/settings.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
+// need md5
+import SettingsUI from '../src/ui/SettingsUI';
+import UserService from './service/UserService';
+import User from './model/User';
+
+const ui = new SettingsUI(document);
+const userService = new UserService();
+let currUser = JSON.parse(localStorage.getItem('currUser'));
```
Comment (yugan2005) [inline src/settings.js]: only saved ID. no need for `JSON.parse`

File: src/settings.js [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+
+ui.saveChangesBtn.addEventListener('click', async () => {
+    // md5 password before checking
```
Comment (yugan2005) [inline src/settings.js]: missing implement

File: src/settings.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+        return;
+    }
+
+    // md5 password before creating user
+    const newUser = new User(ui.newUsername.value, currUser.email, ui.newPassword.value);
+    if (!userService.checkUsername(newUser)) {
+        return;
+    }
+    currUser = await userService.updateUser(currUser.id, newUser);
+    localStorage.setItem('currUser', JSON.stringify(currUser));
```
Comment (yugan2005) [inline src/settings.js]: save only ID

**Review by yugan2005 — CHANGES_REQUESTED** [review]

File: src/settings.js (hunk 21 above)
Comment (tiffanygan) [reply src/settings.js]: `JSON.parse(null) === null`,
`JSON.parse('1') === 1`,
so don't need to change

**Comment by yugan2005** [comment]
LGTM. Merged. Thanks

**Pull request merged**